From eccda3a5f25c2210cd890a45730f1673ea54c324 Mon Sep 17 00:00:00 2001
From: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
Date: Wed, 9 Aug 2023 11:33:17 +0530
Subject: [PATCH] ath12k: KASAN stack out of bounds fix

- Fixed issue Caller has u8 arg and receiver has int argument
- Fixed issue Caller has U8 and receiver has long init argument

U8 -> int
[   33.251666] ==================================================================
[   33.251734] BUG: KASAN: stack-out-of-bounds in ath12k_reg_build_regd+0x998/0xecc [ath12k]
[   33.257811] Write of size 4 at addr ffffff80380c8960 by task swapper/0/0
[   33.266055]
[   33.272823] CPU: 0 PID: 0 Comm: swapper/0 Not tainted 6.1.31 #0
[   33.274301] Hardware name: Qualcomm Technologies, Inc. IPQ5332/RDP442/MI01.3 (DT)
[   33.279947] Call trace:
[   33.287577]  dump_backtrace+0xd4/0xfc
[   33.289837]  show_stack+0x14/0x1c
[   33.293656]  dump_stack_lvl+0x74/0x90
[   33.296954]  print_report+0x9c/0x438
[   33.300600]  kasan_report+0xac/0xb8
[   33.304246]  kasan_check_range+0x28/0x164
[   33.307458]  __asan_storeN+0x1c/0x2c
[   33.311624]  ath12k_reg_build_regd+0x998/0xecc [ath12k]
[   33.315272]  ath12k_wmi_free_tpc_stats_mem+0x9fc/0x3c3c [ath12k]
[   33.320221]  ath12k_htc_rx_completion_handler+0x664/0x67c [ath12k]
[   33.326471]  ath12k_ce_per_engine_service+0x4d8/0x4f8 [ath12k]
[   33.332459]  ath12k_rx_stats_buf_pktlog_process+0x1134/0x25f0 [ath12k]
[   33.338277]  tasklet_action_common.constprop.0+0xf0/0x154

U8->unsigned long
[   33.351509] ==================================================================
[   33.351532] ath12k_ahb soc@0:wifi1@c0000000: Platform name: Qualcomm Technologies, Inc. IPQ5332/RDP442/MI01.3
[   33.351571] BUG: KASAN: stack-out-of-bounds in ath12k_htc_init+0x34c4/0x43fc [ath12k]
[   33.367637] Read of size 8 at addr ffffff8003957ab0 by task kworker/u8:2/106
[   33.375446]
[   33.382562] CPU: 0 PID: 106 Comm: kworker/u8:2 Not tainted 6.1.31 #0
[   33.384039] Hardware name: Qualcomm Technologies, Inc. IPQ5332/RDP442/MI01.3 (DT)
[   33.390381] Workqueue: ath12k_qmi_driver_event ath12k_qmi_qdss_trace_mem_info_send_sync [ath12k]
[   33.397759] Call trace:
[   33.406604]  dump_backtrace+0xd4/0xfc
[   33.408776]  show_stack+0x14/0x1c
[   33.412595]  dump_stack_lvl+0x74/0x90
[   33.415893]  print_report+0x9c/0x438
[   33.419539]  kasan_report+0xac/0xb8
[   33.423184]  kasan_check_range+0x28/0x164
[   33.426398]  __asan_loadN+0x1c/0x2c
[   33.430563]  ath12k_htc_init+0x34c4/0x43fc [ath12k]
[   33.433864]  ath12k_qmi_mlo_global_snapshot_mem_init+0xb4/0x124 [ath12k]
[   33.438728]  ath12k_core_qmi_firmware_ready+0x824/0x1160 [ath12k]
[   33.445672]  ath12k_qmi_qdss_trace_mem_info_send_sync+0xadc/0xd4c [ath12k]

Signed-off-by: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/qmi.c | 5 +++--
 drivers/net/wireless/ath/ath12k/reg.c | 3 ++-
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/qmi.c b/drivers/net/wireless/ath/ath12k/qmi.c
index d06317a..7dd5e1a 100644
--- a/drivers/net/wireless/ath/ath12k/qmi.c
+++ b/drivers/net/wireless/ath/ath12k/qmi.c
@@ -3797,8 +3797,9 @@ ath12k_mgmt_mlo_global_chip_crash_info(struct ath12k_base *ab,
 				       size_t remaining_len,
 				       struct ath12k_host_mlo_glb_chip_crash_info *global_chip_crash_info)
 {
-	int parsed_bytes, len, cur_chip_id;
-	u8 chip_id, valid_chip_bmap;
+	u8 chip_id;
+	s32 len,parsed_bytes;
+	unsigned long valid_chip_bmap, cur_chip_id;
 
 	len = ath12k_mgmt_mlo_global_chip_crash_info_tlv(ab, data, remaining_len,
 							global_chip_crash_info);
-- 
2.17.1

